pandas/core/indexing.py

```diff
@@ -808,6 +808,25 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None) -> None:
         ):
             # GH#38148
             keys = self.obj.columns.union(key, sort=False)
+            diff = Index(key).difference(self.obj.columns, sort=False)
+
+            if len(diff) and com.is_null_slice(orig_key[0]):
+                # e.g. if we are doing df.loc[:, ["A", "B"]] = 7 and "B"
+                #  is a new column, add the new columns with dtype=np.void
+                #  so that later when we go through setitem_single_column
+                #  we will use isetitem. Without this, the reindex_axis
+                #  below would create float64 columns in this example, which
+                #  would successfully hold 7, so we would end up with the wrong
+                #  dtype.
+                arr = np.empty(len(self.obj), dtype=np.void)
```

Comment:
Nit: Is it worth making `arr` 2D first? 


```
arr = np.empty((len(self.obj), len(diff)), dtype=np.void)
df = type(self.obj)(arr, columns=diff, indnex=self.obj.index)
```

Reply:
either way should work.  theres no memory being allocated so everything is really cheap.